gpii/node_modules/flowManager/src/PCPChannel.js

```diff
@@ -0,0 +1,275 @@
+            },
+            // compensate for FLUID-6194
+            backward: "never",
+            forward: {
+                // avoid notifying the client for the init of LifecycleManagerSession, simplifies test and clients
+                excludeSource: "init"
+            }
+        }
+    },
+    modelListeners: {
+        updatePreferences: {
+            path: "{flowManager}.pcpChannel.model.settingControls",
+            includeSource: "PCP",
+            funcName: "gpii.pcpChannel.updatePreferences",
+            args: ["{flowManager}", "{pcpChannel}", "{flowManager}.lifecycleManager", "{change}.value"]
+        },
+        updateContext: {
+            path: "{flowManager}.pcpChannel.model.activeContextName",
+            includeSource: "PCP",
+            funcName: "gpii.pcpChannel.updateContext",
+            args: ["{flowManager}.contextManager", "{pcpChannel}", "{change}.value"]
+        }
+    }
+});
+
+fluid.defaults("gpii.pcpChannel", {
+    gradeNames: ["fluid.modelComponent"],
+    members: {
+        outputBlocked: null
+    },
+    distributeOptions: {
+        // Not {lifecycleManager session} because of FLUID-6193
+        target: "{flowManager lifecycleManager session}.options.gradeNames",
+        record: "gpii.pcpChannel.sessionBinder"
+    },
+    listeners: {
+        "{lifecycleManager}.events.onSessionStop": {
+            namespace: "pcpChannel",
+            funcName: "gpii.pcpChannel.sessionStop",
+            args: "{pcpChannel}"
+        }
+    }
+});
+
+gpii.pcpChannel.updatePreferences = function (flowManager, pcpChannel, lifecycleManager, newPrefs) {
+    fluid.log("Received preferences update from PCP UI of ", newPrefs);
+
+    // Grab current session
+    var session = lifecycleManager.getSession(lifecycleManager.getActiveSessionTokens());
+
+    // tweak payload to be in NP set format and update session
+    var contextualPrefs = {};
+    fluid.each(newPrefs, function (spec, pref) {
+        fluid.set(contextualPrefs, fluid.pathUtil.parseEL(pref), spec.value);
+    });
+
+    var prefcpy = fluid.copy(pcpChannel.model.preferences);
+    fluid.set(prefcpy, ["contexts", session.model.activeContextName, "preferences"], contextualPrefs);
+    session.applier.change(["preferences", "contexts", session.model.activeContextName, "preferences"], contextualPrefs);
+
+    // get new inferred configuration to be applied:
+    var newInferred = gpii.matchMakerFramework.utils.updateInferredConfiguration(session.model.preferences, session.model.matchMakerOutput.inferredConfiguration);
+
+    // full payload from session:
+    var fullPayload = fluid.extend(true, {}, session.model);
+    fluid.set(fullPayload, ["matchMakerOutput", "inferredConfiguration"], newInferred);
+    fullPayload.activeConfiguration = {
+        inferredConfiguration: fluid.extend(true, {}, newInferred[session.model.activeContextName])
+    };
+    lifecycleManager.addLifecycleInstructions(fullPayload);
```

Comment:
I know that the invoker `addLifecycleInstructions` existed before this PR but I am a little confused by it. This invoker causes a side effect to its argument and at least to me, this is not expected given its name (what I would expect is that there is a queue in the `lifecycleManager` where instructions obtained somehow from the payload would be added). Also, it seems that the particular instance of `lifecycleManager` is not used at all in the body of the function.

Reply:
ok, good point! I thought it was clear from the name, but I've now changed it to addLifecycleInstructionToPayload, though a bit long, that should make it clear what it does.. in terms of the particular instance of lifecycle manager, this is being used a couple of lines below in the update call (which depends on the lifecycle managers current state, etc)

Reply:
Sorry for not expressing this good enough - I meant that the `lifecycleManager` instance is not used in the `addLifecycleInstructions` function.  This makes me think whether it won't be better to make this function a utility function (and detach it from the lifecycleManager).